source/common/access_log/access_log_formatter.cc

```diff
@@ -115,28 +100,28 @@ std::string FormatterImpl::format(const Http::HeaderMap& request_headers,
   return log_line;
 }
 
-JsonFormatterImpl::JsonFormatterImpl(std::unordered_map<std::string, std::string>& format_mapping) {
+JsonFormatterImpl::JsonFormatterImpl(std::unordered_map<std::string, std::string>& format_mapping,
+                                     bool preserve_types)
+    : preserve_types_(preserve_types) {
   for (const auto& pair : format_mapping) {
-    auto providers = AccessLogFormatParser::parse(pair.second);
-    json_output_format_.emplace(pair.first, FormatterPtr{new FormatterImpl(pair.second)});
+    json_output_format_.emplace(pair.first, AccessLogFormatParser::parse(pair.second));
   }
 }
 
 std::string JsonFormatterImpl::format(const Http::HeaderMap& request_headers,
                                       const Http::HeaderMap& response_headers,
                                       const Http::HeaderMap& response_trailers,
                                       const StreamInfo::StreamInfo& stream_info) const {
-  const auto output_map = toMap(request_headers, response_headers, response_trailers, stream_info);
-
-  ProtobufWkt::Struct output_struct;
-  for (const auto& pair : output_map) {
-    ProtobufWkt::Value string_value;
-    string_value.set_string_value(pair.second);
-    (*output_struct.mutable_fields())[pair.first] = string_value;
-  }
+  const auto output_struct =
+      toStruct(request_headers, response_headers, response_trailers, stream_info);
 
   std::string log_line;
-  const auto conversion_status = Protobuf::util::MessageToJsonString(output_struct, &log_line);
+  log_line.reserve(256);
+
+  auto print_options = Protobuf::util::JsonPrintOptions();
```

Comment:
can we reuse `MessageUtil::getJsonStringFromMessage`? (you might want to refactor abit for the RELEASE_ASSERT there).